File: src/test/java/net/flyingfishflash/loremlist/unit/core/serialization/UUIDSerializerTest.java
package net.flyingfishflash.loremlist.unit.core.serialization;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import kotlinx.serialization.encoding.Decoder;
import net.flyingfishflash.loremlist.core.exceptions.CoreException;
import net.flyingfishflash.loremlist.core.serialization.UuidSerializer;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpStatus;

class UUIDSerializerTest {

  @Test
  void deserializeThrowsCoreExceptionWithBadRequestStatusForInvalidUuidString() {
    String invalidUuidString = "not-a-uuid";
    Decoder decoder = mock(Decoder.class);
    when(decoder.decodeString()).thenReturn(invalidUuidString);

    assertThatThrownBy(() -> UuidSerializer.INSTANCE.deserialize(decoder))
        .isInstanceOf(CoreException.class)
        .satisfies(
            thrown -> {
              CoreException exception = (CoreException) thrown;
              assertThat(exception.getHttpStatus()).isEqualTo(HttpStatus.BAD_REQUEST);
              assertThat(exception.getMessage()).isEqualTo("Error deserializing UUID.");
              assertThat(exception.getCause()).isInstanceOf(IllegalArgumentException.class);
            });

    verify(decoder).decodeString();
  }
}

File: src/test/kotlin/net/flyingfishflash/loremlist/unit/core/serialization/UUIDSerializerTest.kt
package net.flyingfishflash.loremlist.unit.core.serialization

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.types.shouldBeInstanceOf
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import kotlinx.serialization.encoding.Decoder
import net.flyingfishflash.loremlist.core.exceptions.CoreException
import net.flyingfishflash.loremlist.core.serialization.UuidSerializer
import org.springframework.http.HttpStatus

class UUIDSerializerTest :
  DescribeSpec({
    describe("deserialize") {
      context("when input is an invalid UUID string") {
        it("throws CoreException with BAD_REQUEST status") {
          val invalidUuidString = "not-a-uuid"
          val decoder = mockk<Decoder>()
          every { decoder.decodeString() } returns invalidUuidString
          val exception = shouldThrow<CoreException> { UuidSerializer.deserialize(decoder) }
          exception.httpStatus shouldBe HttpStatus.BAD_REQUEST
          exception.message shouldBe "Error deserializing UUID."
          exception.cause.shouldBeInstanceOf<IllegalArgumentException>()
          verify { decoder.decodeString() }
        }
      }
    }
  })
